File: src-tauri/src/models/step.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum Step {
    Coding {
        id: String,
        title: String,
        concept: String,
        instruction: String,
        hint: Option<String>,
        starter: Option<String>,
        answer: String,
        expected_output: Option<String>,
        validation: ValidationRule,
        encouragement: Option<String>,
    },
    Typing {
        id: String,
        title: String,
        concept: String,
        instruction: String,
        hint: Option<String>,
        target_code: String,
        encouragement: Option<String>,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ValidationRule {
    #[serde(rename = "type")]
    pub rule_type: ValidationType,
    pub value: Option<String>,
    pub pattern: Option<String>,
    pub keywords: Option<Vec<String>>,
    pub exact_match: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ValidationType {
    Contains,
    Regex,
    Exact,
    Ast,
}

File: src/types/step.ts
import type { Difficulty } from './course';

export type StepType = 'coding' | 'typing';

export type ValidationType = 'contains' | 'regex' | 'exact' | 'ast';

export interface ValidationRule {
  type: ValidationType;
  value?: string;
  pattern?: string;
  keywords?: string[];
  exactMatch?: boolean;
}

export interface ExecutionResult {
  success: boolean;
  output: string;
  error: string | null;
  execution_time_ms: number;
  error_type: 'None' | 'CompileError' | 'RuntimeError' | 'Timeout' | 'Unsupported';
}

export interface BaseStep {
  id: string;
  type: StepType;
  title: string;
  concept: string;
  difficulty: Difficulty;
  instruction: string;
  hint?: string;
  encouragement?: string;
  expectedOutput?: string;
}

export interface CodingStep extends BaseStep {
  type: 'coding';
  starter?: string;
  answer: string;
  validation: ValidationRule;
  stdin?: string;
}

export interface TypingStep extends BaseStep {
  type: 'typing';
  targetCode: string;
}

export interface TypingStats {
  wpm: number;
  accuracy: number;
  errors: number;
  totalKeystrokes: number;
  correctKeystrokes: number;
}

export type Step = CodingStep | TypingStep;